src/services/services.ts

```diff
@@ -2044,6 +2059,152 @@ export function createLanguageService(
         return [...semanticDiagnostics, ...declarationDiagnostics];
     }
 
+    function getRegionSemanticDiagnostics(fileName: string, ranges: TextRange[]): RegionDiagnosticsResult | undefined {
+        synchronizeHostData();
+
+        const sourceFile = getValidSourceFile(fileName);
+
+        const options = program.getCompilerOptions();
+        // This is an optimization to avoid computing the nodes in the range if either
+        // we will skip semantic diagnostics for this file or if we already semantic diagnostics for it.
+        if (
+            skipTypeChecking(sourceFile, options, program) ||
+            !canIncludeBindAndCheckDiagnostics(sourceFile, options) ||
+            program.getCachedSemanticDiagnostics(sourceFile)
+        ) {
+            return undefined;
+        }
+
+        const nodes = getNodesForRanges(sourceFile, ranges);
+        if (!nodes) {
+            return undefined;
+        }
+        const checkedSpans = normalizeSpans(nodes.map(node => createTextSpanFromBounds(node.getFullStart(), node.getEnd())));
+        const semanticDiagnostics = program.getSemanticDiagnostics(sourceFile, cancellationToken, nodes);
+        return {
+            diagnostics: semanticDiagnostics.slice(),
+            spans: checkedSpans,
+        };
+    }
+
+    function getNodesForRanges(file: SourceFile, ranges: TextRange[]): Node[] | undefined {

```

Comment:
Maybe this is out of place in our repo, but I almost wonder if it would end up being more performant to make this a generator and pass that down as an interable, avoiding collecting a huge list of nodes that happen to be on screen. But maybe it's not that many.

Reply:
I think that's doable, but I don't know how much that impacts performance. How impactful do you think this change could be?

Reply:
I think it depends on how big these arrays tend to be, and if we ever cancel the diagnostics request. I'm thinking of me looking at a particularly heavy part of the checker and scrolling, where it may turn out to be faster to not have produced a giant list and done a bunch of normalization or something. But, likely not required, just thinking aloud.